### PR TITLE
[Merged by Bors] - feat(NumberTheory/LSeries): even Hurwitz zeta functions (part 1)

### PR DESCRIPTION
**Pull request by loefflerd**
Add the construction of the even part of the Hurwitz zeta function (via FE pairs).

---

The un-completed version, and the relation to the defining Dirichlet series, will come in a later PR, in order to keep each PR in the series a manageable size.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+@[irreducible] def evenKernel (a : UnitAddCircle) (x : ℝ) : ℝ :=
+  (show Function.Periodic
+    (fun ξ : ℝ ↦ rexp (-π * ξ ^ 2 * x) * re (jacobiTheta₂ (ξ * I * x) (I * x))) 1 by
+      intro ξ
+      simp only [ofReal_add, ofReal_one, add_mul, one_mul, jacobiTheta₂_add_left']
+      have : cexp (-↑π * I * ((I * ↑x) + 2 * (↑ξ * I * ↑x))) = rexp (π * (x + 2 * ξ * x)) := by
+        simp only [ofReal_exp, push_cast]
+        ring_nf
+        rw [I_sq]
+        ring_nf
```
Comment (CBirkbeck) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:61]: This is slightly shorter and quicker (although I can't figure out how to get the formatting of the preview to look right).

```suggestion
   have : cexp (-↑π * I * ((I * ↑x) + 2 * (↑ξ * I * ↑x))) = rexp (π * (x + 2 * ξ * x)) := by
        ring_nf
        simp only [I_sq, mul_neg, mul_one, neg_mul, neg_neg, sub_neg_eq_add, ofReal_exp, ofReal_add,
          ofReal_mul, ofReal_ofNat]
```

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+  have : (a' * I * x / (I * x)) = a' := by { field_simp [hx.ne']; ring_nf }
+  rw [this]
+  have : 1 / (-I * (I * x)) ^ (1 / 2 : ℂ) = 1 / ↑(x  ^ (1 / 2 : ℝ)) := by
+    rw [neg_mul, ← mul_assoc, I_mul_I, neg_one_mul, neg_neg,
+        ofReal_cpow hx.le, ofReal_div, ofReal_one, ofReal_ofNat]
+  rw [this]
+  have : -↑π * I * (↑a' * I * ↑x) ^ 2 / (I * ↑x) = - (-π * a' ^ 2 * x) := by
+    rw [mul_pow, mul_pow, I_sq]; field_simp [hx.ne']; ring_nf
+  rw [this, ← mul_assoc, mul_comm (cexp _), mul_assoc _ (cexp _) (cexp _), ← Complex.exp_add,
+      neg_add_self, Complex.exp_zero, mul_one, ofReal_div, ofReal_one]
```
Comment (CBirkbeck) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:149]:
```suggestion
  have h1 : I * ↑(1 / x) = -1 / (I * ↑x) := by
    push_cast; field_simp [hx.ne']; ring_nf; rw [I_sq]; ring_nf
  have h2 : (a' * I * x / (I * x)) = a' := by { field_simp [hx.ne']; ring_nf }
  have h3 : 1 / (-I * (I * x)) ^ (1 / 2 : ℂ) = 1 / ↑(x ^ (1 / 2 : ℝ)) := by
    rw [neg_mul, ← mul_assoc, I_mul_I, neg_one_mul, neg_neg,ofReal_cpow hx.le, ofReal_div, 
      ofReal_one, ofReal_ofNat]
  rw [h1, h2, h3]
  have : -↑π * I * (↑a' * I * ↑x) ^ 2 / (I * ↑x) = - (-π * a' ^ 2 * x) := by
    rw [mul_pow, mul_pow, I_sq]; field_simp [hx.ne']; ring_nf
  rw [this, ← mul_assoc, mul_comm (cexp _), mul_assoc _ (cexp _) (cexp _), ← Complex.exp_add,
      neg_add_self, Complex.exp_zero, mul_one, ofReal_div, ofReal_one]
```

**Review by CBirkbeck — COMMENTED** [review]
This is a far as I got before lunch, will come back later and look at the rest (but it all looks good!)

**Comment by CBirkbeck** [comment]
I just had a quick look at the rest and it all looks good to me :)

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+  (show Function.Periodic
+    (fun ξ : ℝ ↦ rexp (-π * ξ ^ 2 * x) * re (jacobiTheta₂ (ξ * I * x) (I * x))) 1 by
+      intro ξ
+      simp only [ofReal_add, ofReal_one, add_mul, one_mul, jacobiTheta₂_add_left']
+      have : cexp (-↑π * I * ((I * ↑x) + 2 * (↑ξ * I * ↑x))) = rexp (π * (x + 2 * ξ * x)) := by
+        ring_nf
+        simp only [I_sq, mul_neg, mul_one, neg_mul, neg_neg, sub_neg_eq_add, ofReal_exp, ofReal_add,
+          ofReal_mul, ofReal_ofNat]
+      rw [this, re_ofReal_mul, ← mul_assoc, ← Real.exp_add]
+      ring_nf).lift a
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:62]:
```suggestion
      congr
      ring).lift a
```
is a bit faster

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+  · rw [evenKernel_undef _ hx, cosKernel_undef, mul_zero]
+    exact div_nonpos_of_nonneg_of_nonpos zero_le_one hx
+  induction' a using QuotientAddGroup.induction_on' with a
+  rw [← ofReal_inj, ofReal_mul, evenKernel_def, cosKernel_def, jacobiTheta₂_functional_equation]
+  have h1 : I * ↑(1 / x) = -1 / (I * x) := by
+    push_cast
+    rw [← div_div, mul_one_div, div_I, neg_one_mul, neg_neg]
+  have h2 : (a * I * x / (I * x)) = a := by
+    field_simp [hx.ne']
+    ring -- `rw [mul_assoc]` works, but we don't want to rely on the exact output of `field_simp`
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:139]:
```suggestion
  have hx' : I * x ≠ 0 := mul_ne_zero I_ne_zero (ofReal_ne_zero.mpr hx.ne')
  have h2 : (a * I * x / (I * x)) = a := by
    rw [div_eq_iff hx']
    ring
```

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+  have h3 : 1 / (-I * (I * x)) ^ (1 / 2 : ℂ) = 1 / ↑(x ^ (1 / 2 : ℝ)) := by
+    rw [neg_mul, ← mul_assoc, I_mul_I, neg_one_mul, neg_neg,ofReal_cpow hx.le, ofReal_div,
+      ofReal_one, ofReal_ofNat]
+  have h4 : -π * I * (a * I * x) ^ 2 / (I * x) = - (-π * a ^ 2 * x) := by
+    rw [mul_pow, mul_pow, I_sq]; field_simp [hx.ne']; ring_nf
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:144]:
```suggestion
  have h4 : -π * I * (a * I * x) ^ 2 / (I * x) = - (-π * a ^ 2 * x) := by
    rw [mul_pow, mul_pow, I_sq, div_eq_iff hx']
    ring
```
Together with the previous one makes the proof twice as fast. (` field_simp`  is slow here.)

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+
+lemma hasSum_int_evenKernel (a : ℝ) {t : ℝ} (ht : 0 < t) :
+    HasSum (fun n : ℤ ↦ rexp (-π * (n + a) ^ 2 * t)) (evenKernel a t) := by
+  rw [← hasSum_ofReal, evenKernel_def]
+  convert (hasSum_jacobiTheta₂_term _ (by rwa [I_mul_im, ofReal_re])).mul_left _
+    using 2 with n
+  rw [jacobiTheta₂_term, ← Complex.exp_add]
+  push_cast
+  ring_nf
+  simp only [I_sq, mul_neg, neg_mul, mul_one]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:164]:
```suggestion
  have (n : ℤ) : ↑(rexp (-π * (↑n + a) ^ 2 * t)) =
      cexp (-↑π * ↑a ^ 2 * ↑t) * jacobiTheta₂_term n (↑a * I * ↑t) (I * ↑t) := by
    rw [jacobiTheta₂_term, ← Complex.exp_add]
    push_cast
    congr
    ring_nf
    simp only [I_sq, mul_neg, neg_mul, mul_one]
  simp only [this]
  exact (hasSum_jacobiTheta₂_term _ (by rwa [I_mul_im, ofReal_re])).mul_left _
```
Slow ` convert` .

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+
+lemma hasSum_int_cosKernel (a : ℝ) {t : ℝ} (ht : 0 < t) :
+    HasSum (fun n : ℤ ↦ cexp (2 * π * I * a * n) * rexp (-π * n ^ 2 * t)) ↑(cosKernel a t) := by
+  rw [cosKernel_def a t]
+  convert hasSum_jacobiTheta₂_term a (by rwa [I_mul_im, ofReal_re] : 0 < im (I * t)) using 2 with n
+  rw [jacobiTheta₂_term, ofReal_exp, ← Complex.exp_add]
+  push_cast
+  ring_nf
+  simp only [I_sq, mul_neg, neg_mul, mul_one, sub_eq_add_neg]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:173]:
```suggestion
  have (n : ℤ) : cexp (2 * ↑π * I * ↑a * ↑n) * ↑(rexp (-π * ↑n ^ 2 * t)) =
      jacobiTheta₂_term n (↑a) (I * ↑t) := by
    rw [jacobiTheta₂_term, ofReal_exp, ← Complex.exp_add]
    push_cast
    ring_nf
    simp only [I_sq, mul_neg, neg_mul, mul_one, sub_eq_add_neg]
  simp only [this]
  exact hasSum_jacobiTheta₂_term a (by rwa [I_mul_im, ofReal_re] : 0 < im (I * t))
```
Dito.

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+
+/-- Modified version of `hasSum_int_evenKernel` omitting the constant term at `∞`. -/
+lemma hasSum_int_evenKernel₀ (a : ℝ) {t : ℝ} (ht : 0 < t) :
+    HasSum (fun n : ℤ ↦ if n + a = 0 then 0 else rexp (-π * (n + a) ^ 2 * t))
+    (evenKernel a t - if (a : UnitAddCircle) = 0 then 1 else 0) := by
+  simp_rw [AddCircle.coe_eq_zero_iff, zsmul_one]
+  split_ifs with h
+  · obtain ⟨k, rfl⟩ := h
+    simp_rw [← Int.cast_add, Int.cast_eq_zero, add_eq_zero_iff_eq_neg]
+    simpa using hasSum_ite_sub_hasSum (hasSum_int_evenKernel (k : ℝ) ht) (-k)
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:183]:
```suggestion
    simpa only [Int.cast_add, neg_mul, Int.cast_neg, add_left_neg, ne_eq, OfNat.ofNat_ne_zero,
      not_false_eq_true, zero_pow, mul_zero, zero_mul, Real.exp_zero]
      using hasSum_ite_sub_hasSum (hasSum_int_evenKernel (k : ℝ) ht) (-k)
```
is faster, but longer...

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+  · suffices ∀ (n : ℤ), n + a ≠ 0 by simpa [this] using hasSum_int_evenKernel a ht
+    contrapose! h
+    let ⟨n, hn⟩ := h
+    exact ⟨-n, by rwa [Int.cast_neg, neg_eq_iff_add_eq_zero]⟩
+
+lemma hasSum_int_cosKernel₀ (a : ℝ) {t : ℝ} (ht : 0 < t) :
+    HasSum (fun n : ℤ ↦ if n = 0 then 0 else cexp (2 * π * I * a * n) * rexp (-π * n ^ 2 * t))
+    (↑(cosKernel a t) - 1) := by simpa using hasSum_ite_sub_hasSum (hasSum_int_cosKernel a ht) 0
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:191]:
```suggestion
    (↑(cosKernel a t) - 1) := by
  simpa using hasSum_ite_sub_hasSum (hasSum_int_cosKernel a ht) 0
```
is easier to read (one could also squeeze the ` simpa` ...).

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+    HasSum (fun n : ℕ ↦ 2 * Real.cos (2 * π * a * (n + 1)) * rexp (-π * (n + 1) ^ 2 * t))
+    (cosKernel a t - 1) := by
+  rw [← hasSum_ofReal, ofReal_sub, ofReal_one]
+  have := (hasSum_int_cosKernel a ht).sum_nat_of_sum_int
+  rw [← hasSum_nat_add_iff' 1] at this
+  simp_rw [Finset.sum_range_one, Nat.cast_zero, neg_zero, Int.cast_zero, zero_pow two_ne_zero,
+    mul_zero, zero_mul, Complex.exp_zero, Real.exp_zero, ofReal_one, mul_one, Int.cast_neg,
+    Int.cast_natCast, neg_sq, ← add_mul, add_sub_assoc, ← sub_sub, sub_self, zero_sub,
+    ← sub_eq_add_neg, mul_neg] at this
+  convert this with n
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:203]: This ` convert`  is also slow, but the resulting equation is maybe too long to state explicitly...

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+  push_cast
+  rw [Complex.cos, mul_div_cancel₀ _ two_ne_zero]
+  ring_nf
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:206]:
```suggestion
  congr 3 <;> ring
```

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+  obtain ⟨b, _, rfl⟩ := a.eq_coe_Ico
+  obtain ⟨p, hp, hp'⟩ := HurwitzKernelBounds.isBigO_atTop_F_int_zero_sub b
+  refine ⟨p, hp, (EventuallyEq.isBigO ?_).trans hp'⟩
+  filter_upwards [eventually_gt_atTop 0] with t ht
+  simp only [← (hasSum_int_evenKernel b ht).tsum_eq, HurwitzKernelBounds.F_int,
+    HurwitzKernelBounds.f_int, pow_zero, one_mul, Function.Periodic.lift_coe]
+
+/-- The function `cosKernel a - 1` has exponential decay at `+∞`, for any `a`. -/
+lemma isBigO_atTop_cosKernel_sub (a : UnitAddCircle) :
+    ∃ p, 0 < p ∧ IsBigO atTop (cosKernel a · - 1) (fun x ↦ Real.exp (-p * x)) := by
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:225]: Is there a reason to call the positive real whose existence is asserted ` r`  above and ` p`  here?

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+    rw [← isBigO_norm_left] at hv' ⊢
+    conv at hv' =>
+      enter [2, x]; rw [← norm_real, ofReal_sub, apply_ite ((↑) : ℝ → ℂ), ofReal_one, ofReal_zero]
+    exact hv'.trans (isLittleO_exp_neg_mul_rpow_atTop hv _).isBigO
+  g₀ := 1
+  hg_top r := by
+    obtain ⟨p, hp, hp'⟩ := isBigO_atTop_cosKernel_sub a
+    rw [← isBigO_norm_left] at hp' ⊢
+    convert hp'.trans (isLittleO_exp_neg_mul_rpow_atTop hp _).isBigO
+    rw [← norm_real, ofReal_sub, ofReal_one, Function.comp_apply]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:270]:
```suggestion
    have (x : ℝ) : ‖(ofReal' ∘ cosKernel a) x - 1‖ = ‖cosKernel a x - 1‖ := by
      rw [← norm_real, ofReal_sub, ofReal_one, Function.comp_apply]
    simp only [this]
    exact hp'.trans (isLittleO_exp_neg_mul_rpow_atTop hp _).isBigO
```
as usual...

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 14]
```diff
@@ -0,0 +1,386 @@
...
+/-- A `WeakFEPair` structure with `f = evenKernel a` and `g = cosKernel a`. -/
+def hurwitzEvenFEPair (a : UnitAddCircle) : WeakFEPair ℂ where
+  f := ofReal' ∘ evenKernel a
+  g := ofReal' ∘ cosKernel a
+  hf_int := (continuous_ofReal.comp_continuousOn (continuousOn_evenKernel a)).locallyIntegrableOn
+    measurableSet_Ioi
+  hg_int := (continuous_ofReal.comp_continuousOn (continuousOn_cosKernel a)).locallyIntegrableOn
+    measurableSet_Ioi
+  k := 1 / 2
+  hk := by norm_num
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:256]:
```suggestion
  hk := one_half_pos
```

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 15]
```diff
@@ -0,0 +1,386 @@
...
+`1 / s` and `1 / (1 - s)`. -/
+def completedCosZeta₀ (a : UnitAddCircle) (s : ℂ) : ℂ :=
+  ((hurwitzEvenFEPair a).symm.Λ₀ (s / 2)) / 2
+
+lemma completedCosZeta_eq (a : UnitAddCircle) (s : ℂ) :
+    completedCosZeta a s =
+    completedCosZeta₀ a s - 1 / s - (if a = 0 then 1 else 0) / (1 - s) := by
+  rw [completedCosZeta, WeakFEPair.Λ, sub_div, sub_div]
+  congr 1
+  · rw [ completedCosZeta₀, WeakFEPair.symm, hurwitzEvenFEPair, smul_eq_mul, mul_one, div_div,
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:330]:
```suggestion
  · rw [completedCosZeta₀, WeakFEPair.symm, hurwitzEvenFEPair, smul_eq_mul, mul_one, div_div,
```

**Review by MichaelStollBayreuth — APPROVED** [review]
Apart from a few performance issues, this looks good!

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 16]
```diff
@@ -212,7 +212,10 @@ lemma hasSum_nat_cosKernel₀ (a : ℝ) {t : ℝ} (ht : 0 < t) :
     mul_zero, zero_mul, Complex.exp_zero, Real.exp_zero, ofReal_one, mul_one, Int.cast_neg,
     Int.cast_natCast, neg_sq, ← add_mul, add_sub_assoc, ← sub_sub, sub_self, zero_sub,
     ← sub_eq_add_neg, mul_neg] at this
-  convert this with n
+  -- workaround for a slow `convert`
+  -- should this be extracted as a congruence lemma for `HasSum`?
+  have hh {f g : ℕ → ℂ} {z : ℂ} (hf : HasSum f z) (h : g = f) : HasSum g z := h ▸ hf
+  refine hh this <| funext fun n ↦ ?_
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:218]: Now you can use the new congruence lemma here.

Comment (loefflerd) [reply Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean]: Of course. I was just waiting for CI to build all the files in between (I should probably have applied the awaiting-CI tag.) Done now.

**Comment by MichaelStollBayreuth** [comment]
maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**